In [141]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn import linear_model
from sklearn.naive_bayes import GaussianNB

#20151014211444871482

In [2]:
%matplotlib inline

In [38]:
train = pd.read_csv("data/20151014211444871482_dataset.csv", names=["Name", "Personality", "Year", "House", "GPA"])
test = pd.read_csv("data/20151014211444871482_students_2016.csv", names=["Name", "Personality", "Year"])

In [107]:
test["House_manual"] = 0
test["House"] = 0

In [108]:
#train intutions

In [117]:
train.head()

Name Personality  Year       House       GPA
0      Rosmerta         ESTP     1   Slytherin  3.808909
1      Rosmerta         ESTP     2   Slytherin  3.616003
2      Rosmerta         ESTP     3   Slytherin  3.706343
3  Quigley Ogden        ISFJ     1  Hufflepuff  3.297603
4  Quigley Ogden        ISFJ     2   Ravenclaw  3.336324

In [212]:
personality_house_table = pd.pivot_table(train, index=["House", "Personality"], values=["GPA"])

In [317]:
personality_house_table.sort_values(by="GPA", ascending=False).head()

GPA
House      Personality          
Slytherin  ISFJ         3.755265
Gryffindor INFJ         3.753180
Slytherin  ESTP         3.751748
Gryffindor ISTP         3.751370
Hufflepuff ENFJ         3.751088

In [323]:
personality_house_table = pd.pivot_table(train, columns=["House"], index=["Personality"], values=["GPA"])


GPA                               
House       Gryffindor Hufflepuff Ravenclaw Slytherin
Personality                                          
ENFJ          3.249577   3.751088  3.257765  3.248002
ENFP          3.249745   3.749691  3.246165  3.244277
ENTJ          3.247323   3.250110  3.249976  3.249352
ENTP          3.255348   3.249485  3.250067  3.245398
ESFJ          3.245725   3.251780  3.248426  3.248795
ESFP          3.248103   3.249562  3.247741  3.248949
ESTJ          3.251847   3.251557  3.255178  3.250323
ESTP          3.254573   3.246808  3.252724  3.751748
INFJ          3.753180   3.251263  3.247019  3.254218
INFP          3.247323   3.247032  3.749515  3.246010
INTJ          3.248105   3.250570  3.247619  3.247011
INTP          3.250932   3.247916  3.252804  3.249105
ISFJ          3.247557   3.253567  3.248354  3.755265
ISFP          3.251811   3.251426  3.748332  3.247288
ISTJ          3.251183   3.253930  3.248693  3.249923
ISTP          3.751370   3.247214  3.248134  3.251689

best combos: <br>
Slytherin - ISFJ, ESTP <br>
Gryffindor - INFJ, ISTP <br>
Hufflepuff - ENFJ, ENFP <br>
Ravenclaw - INFP, ISFP <br>
All above 3.74 <br>
Other averages around 3.25 <br>


I could now manually create classifiers and assign individuals based on this datatable. I believe these results would be very accurate, but I also want to accomplish this task using machine learning.

In [112]:
#top 8
test.loc[(test.Personality == "ESTP"), "House_manual"] = "Slytherin"
test.loc[(test.Personality == "ISFJ"), "House_manual"] = "Slytherin"
test.loc[(test.Personality == "INFJ"), "House_manual"] = "Gryffindor"
test.loc[(test.Personality == "ISTP"), "House_manual"] = "Gryffindor"
test.loc[(test.Personality == "ENFJ"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "ENFP"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "INFP"), "House_manual"] = "Ravenclaw"
test.loc[(test.Personality == "ISFP"), "House_manual"] = "Ravenclaw"

# Others
test.loc[(test.Personality == "ENTP"), "House_manual"] = "Gryffindor"
test.loc[(test.Personality == "ESTJ"), "House_manual"] = "Ravenclaw"
test.loc[(test.Personality == "ISTJ"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "INTP"), "House_manual"] = "Ravenclaw"
test.loc[(test.Personality == "ESFJ"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "INTJ"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "ENTJ"), "House_manual"] = "Hufflepuff"
test.loc[(test.Personality == "ESFP"), "House_manual"] = "Slytherin"


Also, year doesn't make any difference thus far

In [140]:
# round GPA to 2 decimals:

def truncate(floats):
    return float("{0:.2f}".format(floats))
train.GPA = train.GPA.map(truncate)


In [142]:
gnb = GaussianNB()

In [184]:
input_data = df[["Personality", "GPA"]]
input_data.GPA = input_data.GPA.map(truncate)

In [185]:
input_data.Personality = input_data.Personality.map({'ENFJ': 1, 'ISTJ': 15, 'ESTJ': 7, 'ENFP': 2, 'INTJ': 11, 
 'ENTP': 4, 'ISTP': 16, 'INTP': 12, 'ENTJ': 3, 'ISFJ': 13, 
 'ISFP': 14, 'ESFP': 6, 'INFP': 10, 'INFJ': 9, 'ESTP': 8, 'ESFJ': 5})

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [269]:
df_house = df[["House"]]
house = df_house["House"]

In [270]:
df_house.House = df_house.House.map({"Slytherin": 1, "Hufflepuff": 2, "Ravenclaw": 3, "Gryffindor": 4})

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [242]:
gnb.fit(input_data, house) 

GaussianNB()

In [267]:
for x in range(1,17):
    y = gnb.predict((x, 3.7))
    print(x,y)

1 [2]
2 [2]
3 [2]
4 [2]
5 [2]
6 [2]
7 [1]
8 [1]
9 [1]
10 [3]
11 [3]
12 [3]
13 [4]
14 [4]
15 [4]
16 [4]


Gaussian Naive Bayes works about 60% of the time when graded against my manual classifiers, and it has a low score of .32. Not very happy with it

'ENFJ': 1, 'ISTJ': 15, 'ESTJ': 7, 'ENFP': 2, 'INTJ': 11, <br>
'ENTP': 4, 'ISTP': 16, 'INTP': 12, 'ENTJ': 3, 'ISFJ': 13, <br>
'ISFP': 14, 'ESFP': 6, 'INFP': 10, 'INFJ': 9, 'ESTP': 8, 'ESFJ': 5 <br>
"Slytherin": 1, "Hufflepuff": 2, "Ravenclaw": 3, "Gryffindor": 4 <br>

"ESTP" 8 = 1 "Slytherin" <br>
"ISFJ" 13= 1 "Slytherin"<br>
"INFJ" 9 = 4 "Gryffindor"<br>
"ISTP" 16 = 4 "Gryffindor"<br>
"ENFJ" 1 = 2 "Hufflepuff"<br>
"ENFP" 2 = 2 "Hufflepuff"<br>
"INFP" 10 = 3 "Ravenclaw"<br>
"ISFP" 14 = 3 "Ravenclaw"<br>
<br>
"ENTP" 4 = 4 "Gryffindor"<br>
"ESTJ" 7 = 3 "Ravenclaw"<br>
"ISTJ" 15 = 2 "Hufflepuff"<br>
"INTP" 12 = 3 "Ravenclaw"<br>
"ESFJ" 5 = 2 "Hufflepuff"<br>
"INTJ" 11 = 2 "Hufflepuff"<br>
"ENTJ 3 = 2 "Hufflepuff"<br>
"ESFP" 6 = 1 "Slytherin"<br>

In [249]:
gnb.score(input_data, house)

0.32183670485993843

In [257]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB 


In [258]:
nb = MultinomialNB()
bnb = BernoulliNB()

In [259]:
nb.fit(input_data, house) 
bnb.fit(input_data, house) 

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)


BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

In [262]:
for x in range(1,17):
    y = nb.predict((x, 3.7))
    print(x,y)

1 [2]
2 [2]
3 [2]
4 [2]
5 [2]
6 [2]
7 [2]
8 [2]
9 [4]
10 [4]
11 [4]
12 [4]
13 [4]
14 [4]
15 [4]
16 [4]


Multinomial Naive Bayes does not work as it outputs only two classes

In [260]:
for x in range(1,17):
    y = bnb.predict((x, 3.7))
    print(x,y)

1 [4]
2 [4]
3 [4]
4 [4]
5 [4]
6 [4]
7 [4]
8 [4]
9 [4]
10 [4]
11 [4]
12 [4]
13 [4]
14 [4]
15 [4]
16 [4]


Bernoulli Naive Bayes also does not work as it outputs only one classe

Now I'm going to make the GPAs a discrete rather than continuous variable by putting them into buckets.

In [284]:
input_d)

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [290]:
input_data = df[["Personality", "GPA"]]
input_data.Personality = input_data.Personality.map({'ENFJ': 1, 'ISTJ': 15, 'ESTJ': 7, 'ENFP': 2, 'INTJ': 11, 
 'ENTP': 4, 'ISTP': 16, 'INTP': 12, 'ENTJ': 3, 'ISFJ': 13, 
 'ISFP': 14, 'ESFP': 6, 'INFP': 10, 'INFJ': 9, 'ESTP': 8, 'ESFJ': 5})
input_data.GPA = input_data.GPA.map(truncate)

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [423]:
df_house = df[["House"]]
house = df_house["House"]

In [315]:
df_house.House = df_house.House.map({"Slytherin": 1, "Hufflepuff": 2, "Ravenclaw": 3, "Gryffindor": 4})
house = df_house.House

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [416]:
personality_house_table.head()

GPA                               
House       Gryffindor Hufflepuff Ravenclaw Slytherin
Personality                                          
ENFJ          3.249577   3.751088  3.257765  3.248002
ENFP          3.249745   3.749691  3.246165  3.244277
ENTJ          3.247323   3.250110  3.249976  3.249352
ENTP          3.255348   3.249485  3.250067  3.245398
ESFJ          3.245725   3.251780  3.248426  3.248795

In [411]:
gryffindor = personality_house_table['GPA']['Gryffindor']
hufflepuff = personality_house_table['GPA']['Hufflepuff']
ravenclaw = personality_house_table['GPA']['Ravenclaw']
slytherin = personality_house_table['GPA']['Slytherin']

In [412]:
gryffindor = pd.DataFrame(gryffindor)
gryffindor['House'] = 'Gryffindor'
gryffindor.reset_index(inplace=True)
gryffindor = gryffindor.rename(columns={"Gryffindor": "GPA"})

hufflepuff = pd.DataFrame(hufflepuff)
hufflepuff['House'] = 'Hufflepuff'
hufflepuff.reset_index(inplace=True)
hufflepuff = hufflepuff.rename(columns={"Hufflepuff": "GPA"})

ravenclaw = pd.DataFrame(ravenclaw)
ravenclaw['House'] = 'Ravenclaw'
ravenclaw.reset_index(inplace=True)
ravenclaw = ravenclaw.rename(columns={"Ravenclaw": "GPA"})

slytherin = pd.DataFrame(slytherin)
slytherin['House'] = 'Slytherin'
slytherin.reset_index(inplace=True)
slytherin = slytherin.rename(columns={"Slytherin": "GPA"})


In [430]:
houses = [gryffindor, hufflepuff, ravenclaw, slytherin]
pivot_data = pd.concat(houses)
pivot_data.index = range(0,64)

In [465]:
input_data = pivot_data[["Personality", "GPA"]]
input_data.GPA = input_data.GPA.map(truncate)
input_data.Personality = input_data.Personality.map({'ENFJ': 1, 'ISTJ': 15, 'ESTJ': 7, 'ENFP': 2, 'INTJ': 11, 
 'ENTP': 4, 'ISTP': 16, 'INTP': 12, 'ENTJ': 3, 'ISFJ': 13, 
 'ISFP': 14, 'ESFP': 6, 'INFP': 10, 'INFJ': 9, 'ESTP': 8, 'ESFJ': 5})

/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [466]:
input_data

Personality   GPA
0             1  3.25
1             2  3.25
2             3  3.25
3             4  3.26
4             5  3.25
5             6  3.25
6             7  3.25
7             8  3.25
8             9  3.75
9            10  3.25
10           11  3.25
11           12  3.25
12           13  3.25
13           14  3.25
14           15  3.25
15           16  3.75
16            1  3.75
17            2  3.75
18            3  3.25
19            4  3.25
20            5  3.25
21            6  3.25
22            7  3.25
23            8  3.25
24            9  3.25
25           10  3.25
26           11  3.25
27           12  3.25
28           13  3.25
29           14  3.25
..          ...   ...
34            3  3.25
35            4  3.25
36            5  3.25
37            6  3.25
38            7  3.26
39            8  3.25
40            9  3.25
41           10  3.75
42           11  3.25
43           12  3.25
44           13  3.25
45           14  3.75
46           15  3.25
47           16  3.25
48            1  3.25
49            2  3.24
50            3  3.25
51            4  3.25
52            5  3.25
53            6  3.25
54            7  3.25
55            8  3.75
56            9  3.25
57           10  3.25
58           11  3.25
59           12  3.25
60           13  3.76
61           14  3.25
62           15  3.25
63           16  3.25

[64 rows x 2 columns]

In [467]:
df_house = pivot_data[["House"]]
df_house.House = df_house.House.map({"Slytherin": 1, "Hufflepuff": 2, "Ravenclaw": 3, "Gryffindor": 4})
house = df_house["House"]


/Users/MattKasle/Matt/.direnv/python-3.5.0/lib/python3.5/site-packages/pandas/core/generic.py:2273: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [470]:
gnb = GaussianNB()
gnb.fit(input_data, house) 


GaussianNB()

In [471]:
for x in range(1,17):
    y = gnb.predict((x, 3.0))
    print(x,y)

1 [1]
2 [1]
3 [1]
4 [1]
5 [1]
6 [1]
7 [1]
8 [1]
9 [1]
10 [1]
11 [1]
12 [1]
13 [1]
14 [1]
15 [1]
16 [1]


Not working with any type of Bayes :(